### PR TITLE
Allow action parameters to be controllable

### LINKED ISSUE #22068: Make action parameters also controllable
## Reason
It is desired to change certain parameters in an input file during the calculation. This can be achieved with controllable parameter capability. However, currently, only objects derived from `MooseObject` can have controllable parameters. A large portion of input syntax with `Action`s cannot have controllable parameters. A parameter in an action might be used for constructing bunch of moose objects, it will be up to action developers to connect the controllable parameter in an action with the parameters of all the depending moose objects, which has been a capability in the input parameter warehouse.

## Design
To achieve this, we will first need to make the parameters of various actions held by the `InputParameterWarehouse` like the moose objects but instead of the action itself, and change the constructor interface from
```
  Action(InputParameters parameters);
```
to
```
  Action(const InputParameters & parameters);
```
This will be a relative large change in interface and will affect all applications.

## Impact
A new capability.

### PR DESCRIPTION
**Pull request by YaqiWang**
Close #22068.

This is still WIP. I will need to prepare patches for all modules/applications that have actions due to the interface change of the action constructor. This should pass moose framework tests though.

### CODE REVIEW

File: modules/tensor_mechanics/src/actions/CohesiveZoneActionBase.C
Comment (YaqiWang) [inline modules/tensor_mechanics/src/actions/CohesiveZoneActionBase.C:95]: Now the parameter is held in the warehouse and the action only owns a const reference of it. To get access to that input parameters, we have to inquire the warehouse. I feel you can use an action of action to replace the current syntax used by this action and to construct the full input parameters and then let the action factory create this action.

File: modules/tensor_mechanics/src/actions/CohesiveZoneActionBase.C
Comment (dschwen) [inline modules/tensor_mechanics/src/actions/CohesiveZoneActionBase.C:96]: Why can't this be in the base class constructor? And why do the parameters in the warehouse have to be modifiable for anyone this to work?

Comment (dschwen) [reply modules/tensor_mechanics/src/actions/CohesiveZoneActionBase.C:96]: and what is "action of action"?

Comment (YaqiWang) [reply modules/tensor_mechanics/src/actions/CohesiveZoneActionBase.C:96]: The base class can construct a non-const reference. Not get the second question.

"action of action" means a meta action that can call `ActionFactory::create` to add a new action. In there, the action can populate necessary parameters of the action to be created and then create the action. The action to be created will not register any input syntax. Creating the action can happen at `meta_action` task which is at the beginning of execution of all actions.

**Comment by YaqiWang** [comment]
@lindsayad I tried to remove that restriction on `ptr->allowCopy(false)` but cannot make it work. Once that function is called on a parameter, the parameter will no longer be able to be copied, which is what `ActionWarehouse::printInputFile` is doing. `allowCopy` is a private function of `InputParameters`, so we cannot revert it either. I tried to make the `SyntaxTree` to not create copies of `InputParameters`, it almost works but failed at `Parser::buildFullTree` for some reason. I attach the patch I attempted for your info. I guess we will have to allow parameter to be copied for actions.
[syntax_tree.txt](https://github.com/idaholab/moose/files/9551210/syntax_tree.txt)

**Comment by YaqiWang** [comment]
@amccaskey this is ready for review. Now only app tests are failing. One thing related with @dschwen 's comment is that I removed the non-const `InputParameters & parameters()` in `Action.h`. I think having that in `Action` is possibly not a good idea. But I am not against to put it back. I did have the implementation of that function already in `GlobalParamsAction`.

I will need to patch applications due to the constructor interface change. It should be quite straight-forward like what I did in modules.

**Comment by lindsayad** [comment]
@permcody are you on-board with this concept or do you have major objections?

**Comment by permcody** [comment]
I don't love the idea of this, but I also realize that there are several framework team members that do and are supportive so I'm not going to hold you all back. Also, I'm not into the code enough to understand any side effects. There will likely be some as this inserts a pretty big back door into the code (outside of the Control system) if I'm reading this all right. I won't hold this up.

**Comment by friedmud** [comment]
Despite some reservations - I do think this is shaping up well.

One question: is there anything we can do about backwards compatibility.  This will break anyone with a custom Action (right?)

File: framework/include/utils/InputParameterWarehouse.h
Comment (lindsayad) [inline framework/include/utils/InputParameterWarehouse.h:181]: Why is this necessary?

File: framework/include/utils/InputParameters.h
Comment (lindsayad) [inline framework/include/utils/InputParameters.h:1018]: and this?

Comment (YaqiWang) [reply framework/include/utils/InputParameters.h:1018]: This one is for that printInputFile function.

File: framework/include/utils/InputParameterWarehouse.h
Comment (YaqiWang) [reply framework/include/utils/InputParameterWarehouse.h:181]: This is for calling `addInputParameters` function similar as `Factory::create`.

**Comment by YaqiWang** [comment]
> One question: is there anything we can do about backwards compatibility. This will break anyone with a custom Action (right?)

This is a new capability. It breaks applications only because the Action constructor interface change. Few places rely on the non-const references to the input parameters, which is a little harder to obtain, but still can though the InputParametersWarehouse.

File: framework/doc/content/syntax/Controls/index.md
Comment (joshuahansel) [inline framework/doc/content/syntax/Controls/index.md:96]: [Actions]

File: framework/doc/content/syntax/Controls/index.md
Comment (joshuahansel) [inline framework/doc/content/syntax/Controls/index.md:97]: You shouldn't need to manually name the page "Creating a Controllable Parameter" - the section name should automatically be the default link text, right?

File: framework/src/actions/Action.C
Comment (joshuahansel) [inline framework/src/actions/Action.C:189]: Can you pass in `object_params` instead of `object_type`? This is what we do in `Component`s in THM. Then you can get the base type with `object_params.get<std::string>("_moose_base")`.

File: framework/src/actions/Action.C
Comment (joshuahansel) [inline framework/src/actions/Action.C:49]: Is this documented somewhere?

**Review by joshuahansel — CHANGES_REQUESTED** [review]

**Comment by joshuahansel** [comment]
> One question: is there anything we can do about backwards compatibility. This will break anyone with a custom Action (right?)

I was wondering the same - if there's some kind of deprecation procedure we can do by overloading the constructor instead of replacing it, and having some deprecation message in the old constructor.

File: framework/doc/content/syntax/Controls/index.md
Comment (YaqiWang) [reply framework/doc/content/syntax/Controls/index.md:97]: I will have to run the server to double check. I was unsure whether `[syntax/Controls/index.md#sec:control-param]` is enough. I initially had `[sec:control-param]` but for some reason that does not work.

File: framework/src/actions/Action.C
Comment (YaqiWang) [reply framework/src/actions/Action.C:49]: In the parameter list. The `Controls/index.md` unfortunately does not mention this. I can add something there.

**Comment by YaqiWang** [comment]
> > One question: is there anything we can do about backwards compatibility. This will break anyone with a custom Action (right?)
> 
> I was wondering the same - if there's some kind of deprecation procedure we can do by overloading the constructor instead of replacing it, and having some deprecation message in the old constructor.

Possibly not, because the old interface is passing in the input parameters by value. With the changes in `ActionFactory`, the parameter is not allowed to be copied thus I think passing in values will cause an error.

File: framework/src/actions/Action.C
Comment (YaqiWang) [reply framework/src/actions/Action.C:189]: I think this is more straight-forward to developers. At the place this function is called, developers know exactly the object type, name and parameter. It will take them some thinking to pass in the MooseObjectName or MooseObjectParameterName.

Comment (YaqiWang) [reply framework/src/actions/Action.C:189]: Oh, I see what you mean. Yeah, MOOSE base is a bit difficult for developers to understand. Ideally here we pass in the object type, but in reality here it is a MOOSE base name. Is there a function somewhere we can use to get the base from a type, like `giveMeBase("Diffusion")` returning `Kernel`?

Comment (YaqiWang) [reply framework/src/actions/Action.C:189]: I will just pass in the parameter name. If we do this, we do not need the object name as well because it can be obtained from `_object_name`.

File: framework/src/actions/ActionFactory.C
Comment (lindsayad) [inline framework/src/actions/ActionFactory.C:56]: Can we do this towards the end of `Parser::walkRaw` in the vicinity of where we modify `control_tags` for the object params of moose object actions? In this location it just feels out of the blue and a developer who comes along will have to go search through the code base to understand why there is special treatment given to non-moose-object-actions

File: framework/src/actions/ActionFactory.C
Comment (lindsayad) [inline framework/src/actions/ActionFactory.C:61]: Can we rename `params` and/or `parameters` to indicate more clearly what their difference is? Maybe `action_params` and `incoming_parser_params` respectively?

File: framework/include/utils/InputParameters.h
Comment (lindsayad) [reply framework/include/utils/InputParameters.h:1018]:
```suggestion
  friend void ActionWarehouse::printInputFile(std::ostream & out);
```
Let's make it specific to where we need it then

File: framework/doc/content/syntax/Controls/index.md
Comment (lindsayad) [inline framework/doc/content/syntax/Controls/index.md:155]:
```suggestion
         block=Controls
```

**Review by lindsayad — CHANGES_REQUESTED** [review]
overall looking pretty good to me

File: framework/src/actions/ActionFactory.C
Comment (YaqiWang) [reply framework/src/actions/ActionFactory.C:56]: There we do not have access to the parameters in the warehouse. For moose object actions we can do that because we are modifying the parameters of the objects but not the actions. I am guessing it does not hurt to remove the if statement here.

Comment (lindsayad) [reply framework/src/actions/ActionFactory.C:56]: You are operating on the exact same parameters object here. This code in `Parser::walkRaw`:
```c++
    // Create the Action
    std::shared_ptr<Action> action_obj =
        _action_factory.create(it->second._action, MooseUtils::shortName(curr_identifier), params);
```

Comment (YaqiWang) [reply framework/src/actions/ActionFactory.C:56]: right. Will do. We can add this before calling input warehouse add input param function.

File: modules/tensor_mechanics/src/actions/CohesiveZoneActionBase.C
Comment (YaqiWang) [reply modules/tensor_mechanics/src/actions/CohesiveZoneActionBase.C:97]: @dschwen are you ok with this or you prefer me to add the non-const function to get the action parameters in the base class?

File: framework/doc/content/syntax/Controls/index.md
Comment (YaqiWang) [reply framework/doc/content/syntax/Controls/index.md:155]: @lindsayad if we make friend more specific as suggested, I am getting a compiling error
```
error: use of undeclared identifier 'ActionWarehouse'; did you mean 'OutputWarehouse'?
```
Unfortunately we cannot include `ActionWarehouse.h` in `InputParameters.h` because of the cyclic including I think. Any suggestions?

**Comment by YaqiWang** [comment]
This command is pretty handy for patching applications:
```
grep -rl "(InputParameters param" [path to search] --include="*.h" | xargs sed -i '' "s/(InputParameters param/(const InputParameters \& param/g"
```

File: framework/include/utils/InputParameters.h
Comment (lindsayad) [reply framework/include/utils/InputParameters.h:1019]: > Unfortunately we cannot include ActionWarehouse.h in InputParameters.h because of the cyclic including I think. Any suggestions?

Now that you are storing a reference to the input parameters in `Action`, you should be able to change the `#include "InputParameters.h` in that file to a forward declaration which would allow you to include the action warehouse header in `InputParameters.h`

**Comment by lindsayad** [comment]
@joshuahansel can you check and see whether your conversations have been resolved?

File: framework/include/utils/InputParameters.h
Comment (YaqiWang) [reply framework/include/utils/InputParameters.h:1019]: Let me try. But I've found there are bunch of other places indirectly included by Action.h having includes of InputParameters.h.

Comment (YaqiWang) [reply framework/include/utils/InputParameters.h:1019]: Unfortunately, there are some template functions in `Action.h` relying on `InputParameters.h`... Same for the `Registry.h`, included by `Action.h`. I remove the unnecessary include of `Registry.h` in `Action.h` and few other indirect includes, but we cannot remove the include in `Action.h` itself. Also we cannot remove the include of `Action.h` in `ActionWarehouse.h`. So I think the possible solution could be move this function out of `ActionWarehouse`, for example in the Parser. But that requires the action warehouse to provide a public function to expose all its actions, which may not be a good idea either. Another solution is to let the `InputParameterWarehouse` to do the print, but that could be other implications. I also feel maybe we should just let parser do the print.

So I suggest we keep this. Are you convinced? I will push what I have on some attempts of reducing header includes in the last commit.

File: framework/doc/content/syntax/Controls/index.md
Comment (joshuahansel) [reply framework/doc/content/syntax/Controls/index.md]: `[sec:control-param]` won't be enough, but `[syntax/Controls/index.md#sec:control-param]` should be.

File: framework/src/actions/Action.C
Comment (joshuahansel) [reply framework/src/actions/Action.C:49]: Yes, I realize there is a parameter description :) After reading that description, I still don't know how to use it. I looked in `Controls/index.md` as well and couldn't find it, so that'd be great if you could document that.

File: framework/include/utils/InputParameters.h
Comment (lindsayad) [reply framework/include/utils/InputParameters.h:1019]: fine

File: modules/tensor_mechanics/src/actions/CohesiveZoneActionBase.C
Comment (lindsayad) [reply modules/tensor_mechanics/src/actions/CohesiveZoneActionBase.C:95]: Seems like a duplicate thread of https://github.com/idaholab/moose/pull/22069/files#r968702615

File: modules/tensor_mechanics/src/actions/CohesiveZoneActionBase.C
Comment (lindsayad) [reply modules/tensor_mechanics/src/actions/CohesiveZoneActionBase.C:97]: I don't want to see
```c++
InputParameters & parameters();
```
added back to `Action`. I would rather you just `const_cast<InputParameters &>(_pars)` here because a `const_cast` is an indication that a developer is doing something that we'd rather not have them do. That you can do the above two lines and get a non-const reference to the parameters is unfortunate IMO and I don't want it advertised.

**Comment by lindsayad** [comment]
@YaqiWang can you start a list of the apps that need to be fixed for CI and your status on fixing them?

**Comment by lindsayad** [comment]
That list will need to be complete before we can consider merging this. The alternative is to think about how we would make this backwards compatible (and admittedly I think that would be pretty tough to do)

File: modules/tensor_mechanics/src/actions/CohesiveZoneActionBase.C
Comment (YaqiWang) [reply modules/tensor_mechanics/src/actions/CohesiveZoneActionBase.C:97]: ok.

**Comment by YaqiWang** [comment]
Apps need to be patched:

- [x] Griffin
- [ ] Bison
- [ ] SAM
- [ ] RELAP-7
- [ ] Subchannel
- [ ] Pronghorn
- [ ] Okami
- [ ] Pika
- [ ] Zapdos
- [ ] Moltres
- [ ] Ferret
- [ ] Raccoon
- [ ] Isopod

I think most of them just need to update the action constructors, which works with the current moose, so they can be merged first (of cause with the checks of the latest interface here). I have done the patch for Griffin. Other than SAM, RELAP-7 and Subchannel, I can do all of them. But if someone can help, it is greatly appreciated. Bluecrab, direwolf, and sabertooth will need to have submodule update later.

File: modules/tensor_mechanics/src/actions/CohesiveZoneActionBase.C
Comment (lindsayad) [reply modules/tensor_mechanics/src/actions/CohesiveZoneActionBase.C:97]: @dschwen can we resolve this?

**Comment by YaqiWang** [comment]
@lindsayad shall we merge this and start patching Griffin/Bison/Moltres/SAM and then wrapping apps?

**Review by lindsayad — APPROVED** [review]

**Comment by lindsayad** [comment]
@joshuahansel do you still have requested changes?

**Review by joshuahansel — APPROVED** [review]

**Comment by joshuahansel** [comment]
Nope, just approved.

**Pull request merged**
